### DISCUSSION TITLE
Some performance related observations about ReadOnlyBase

### DISCUSSION THREAD

**Discussion by Chicagoan2016** [MarimerLLC/csla#3428] [Questions]
I am trying to find out performance bottlenecks in our application and two things stand out in dottrace (JetBrains) tool.
One is the VerifyAuthoriztionCache and other is CanReadProperty, both from Csla.ReadOnlyBase.

In our Readonly objects we are not using any authorization rules, meaning the method
'AddObjectAuthorizationRules()' in our parent readonly list is empty and same is the case for 
'AddBusinessRules' in the child read only objects.

Kind Regards

**Comment by rockfordlhotka** [comment]
I'm sure `VerifyAuthorizationCache` is called frequently - but it shouldn't be slow, as it doesn't do much:

```c#
    private void VerifyAuthorizationCache()
    {
      if (_readResultCache == null)
        _readResultCache = new ConcurrentDictionary<string, bool>();
      if (_executeResultCache == null)
        _executeResultCache = new ConcurrentDictionary<string, bool>();
      if (!ReferenceEquals(ApplicationContext.User, _lastPrincipal))
      {
        // the principal has changed - reset the cache
        _readResultCache.Clear();
        _executeResultCache.Clear();
        _lastPrincipal = ApplicationContext.User;
      }
    }
```

`CanReadProperty` does a dictionary lookup each time, and that's not free, but should be pretty fast. The dictionary (in your case) will be a list of `true` values since there are no rules. The cache exists to avoid re-checking the rules every time - the cache is flushed only when the current user identity changes.

**Reply by Chicagoan2016** [reply]
@rockfordlhotka Thank you Sir, maybe in next version of Csla, we should have an option to turn these off?

**Reply by rockfordlhotka** [reply]
At a per-type level, or for the whole app?

I'd think it would need to be per-type.

**Reply by Chicagoan2016** [reply]
@rockfordlhotka  I agree, it should be per type.

**Reply by rockfordlhotka** [reply]
https://github.com/MarimerLLC/csla/issues/3430
